File: src/main/java/simpledb/record/RecordPage.java
package simpledb.record;

import simpledb.file.BlockId;
import simpledb.tx.Transaction;

import static java.sql.Types.INTEGER;

public class RecordPage {
    public static final int EMPTY = 0, USED = 1;
    private Transaction tx;
    private BlockId blk;
    private Layout layout;
    public RecordPage(Transaction tx, BlockId blk, Layout layout) {
        this.tx = tx;
        this.blk = blk;
        this.layout = layout;
        tx.pin(blk);
    }

    public int getInt(int slot, String fldname) {
        int fldpos = offset(slot) + layout.offset(fldname);
        return tx.getInt(blk, fldpos);
    }

    public String getString(int slot, String fldname) {
        int fldpos = offset(slot) + layout.offset(fldname);
        return tx.getString(blk, fldpos);
    }

    public void setInt(int slot, String fldname, int val) {
        int fldpos = offset(slot) + layout.offset(fldname);
        tx.setInt(blk, fldpos, val, true);;
    }

    public void setString(int slot, String fldname, String val) {
        int fldpos = offset(slot) + layout.offset(fldname);
        tx.setString(blk, fldpos, val, true);;
    }

    public void delete(int slot) {
        setFlag(slot, EMPTY);
    }

    public void format() {
        int slot = 0;
        while (isValidSlot(slot)) {
            tx.setInt(blk, offset(slot), EMPTY, false);
            Schema sch = layout.schema();
            for (String fldname: sch.fields()) {
                int fldpos = offset(slot) + layout.offset(fldname);
                if (sch.type(fldname) == INTEGER) {
                    tx.setInt(blk, fldpos, 0, false);
                } else {
                    tx.setString(blk, fldpos, "", false);
                }
            }
            slot++;
        }
    }

    public int nextAfter(int slot) {
        return searchAfter(slot, USED);
    }

    public int insertAfter(int slot) {
        int newslot = searchAfter(slot, EMPTY);
        if (newslot >= 0)
            setFlag(newslot, USED);
        return newslot;
    }

    public BlockId block() {
        return blk;
    }

    private void setFlag(int slot, int flag) {
        tx.setInt(blk, offset(slot), flag, true);
    }

    private int searchAfter(int slot, int used) {
        slot++;
        while(isValidSlot(slot)) {
            if (tx.getInt(blk, offset(slot)) == used) {
                return slot;
            }
            slot++;
        }
        return -1;
    }

    private boolean isValidSlot(int slot) {
        return offset(slot+1) <= tx.blockSize();
    }

    private int offset(int slot) {
        return slot * layout.slotSize();
    }

}

File: src/main/kotlin/simpledb/record/rowlock/RecordPage.kt
package simpledb.record.rowlock

import simpledb.file.BlockId
import simpledb.record.Layout
import simpledb.tx.rowlock.TransactionImpl
import java.sql.Types

class RecordPage(private val tx: TransactionImpl, private val blk: BlockId, private val layout: Layout) {
    init {
        tx.pin(blk)
    }

    fun getInt(slot: Int, fldName: String): Int {
        val fldPos = offset(slot) + layout.offset(fldName)
        return tx.getInt(blk, fldPos)
    }

    fun getString(slot: Int, fldName: String): String {
        val fldPos = offset(slot) + layout.offset(fldName)
        return tx.getString(blk, fldPos)
    }

    fun setInt(slot: Int, fldName: String, value: Int) {
        tx.wLatchPage(blk)
        try {
            tx.lockExclusive(blk, slot)
            val fldPos = offset(slot) + layout.offset(fldName)
            tx.setInt(blk, fldPos, value, true)
        } finally {
            tx.wUnlatchPage(blk)
        }
    }

    fun setString(slot: Int, fldName: String, value: String) {
        tx.wLatchPage(blk)
        try {
            tx.lockExclusive(blk, slot)
            val fldPos = offset(slot) + layout.offset(fldName)
            tx.setString(blk, fldPos, value, true)
        } finally {
            tx.wUnlatchPage(blk)
        }
    }

    fun delete(slot: Int) {
        setFlag(slot, EMPTY)
    }

    fun format() {
        var slot = 0
        while (isValidSlot(slot)) {
            tx.setInt(blk, offset(slot), EMPTY, false)
            val sch = layout.schema()
            for (fldName in sch.fields()) {
                val fldPos = offset(slot) + layout.offset(fldName)
                if (sch.type(fldName) == Types.INTEGER) {
                    tx.setInt(blk, fldPos, 0, false)
                } else {
                    tx.setString(blk, fldPos, "", false)
                }
            }
            slot++
        }
    }

    fun nextAfter(slot: Int): Int {
        return searchAfter(slot, USED)
    }

    fun insertAfter(slot: Int): Int {
        val newSlot = searchAfter(slot, EMPTY)
        if (newSlot >= 0) setFlag(newSlot, USED)
        return newSlot
    }

    fun block(): BlockId {
        return blk
    }

    private fun setFlag(slot: Int, flag: Int) {
        tx.setInt(blk, offset(slot), flag, true)
    }

    private fun searchAfter(orgSlot: Int, used: Int): Int {
        var slot = orgSlot
        slot++
        while (isValidSlot(slot)) {
            tx.rLatchPage(blk)
            try {
                tx.lockShared(blk, slot)
                if (tx.getInt(blk, offset(slot)) == used) {
                    return slot
                }
            } finally {
                tx.rUnlatchPage(blk)
            }
            slot++
        }
        return -1
    }

    private fun isValidSlot(slot: Int): Boolean {
        return offset(slot + 1) <= tx.blockSize()
    }

    private fun offset(slot: Int): Int {
        return slot * layout.slotSize()
    }

    companion object {
        const val EMPTY = 0
        const val USED = 1
    }
}
